File: AdvancedMap.Objective-C/AdvancedMap/GeocodingView.m
//
//  GeocodingView.m
//  AdvancedMap.Objective-C
//
//  Created by Aare Undo on 19/09/2017.
//  Copyright © 2017 Nutiteq. All rights reserved.
//

#import "GeocodingView.h"

@implementation GeocodingView

- (id) init {
    self = [super init];

    self.IDENTIFIER = @"result_table_id";
    
    self.darkTransparentGray = [UIColor colorWithRed:50 / 255.0 green:50 / 255.0 blue:50 / 255.0 alpha:200 / 255.0];
    self.lightTransparentGray = [UIColor colorWithRed:50 / 255.0 green:50 / 255.0 blue:50 / 255.0 alpha:120 / 255.0];
    self.transparent = [UIColor colorWithRed:0 / 255.0 green:0 / 255.0 blue:0 / 255.0 alpha:0 / 255.0];
    
    self.font = [UIFont fontWithName:@"HelveticaNeue" size:15];
    
    self.addresses = [[NSMutableArray alloc]init];
    
    self.inputField = [[UITextField alloc]init];
    [self.inputField setTextColor:UIColor.whiteColor];
    [self.inputField setBackgroundColor:self.darkTransparentGray];
    [self.inputField setAutocorrectionType:UITextAutocorrectionTypeNo];
    [self.inputField setFont:self.font];
    // Text padding
    UIView *view = [[UIView alloc]initWithFrame:CGRectMake(0, 0, 10, 20)];
    [self.inputField setLeftView:view];
    [self.inputField setLeftViewMode:UITextFieldViewModeAlways];
    [self addSubview:_inputField];
    
    self.resultTable = [[UITableView alloc]init];
    [self.resultTable registerClass:[UITableViewCell class] forCellReuseIdentifier:self.IDENTIFIER];
    [self.resultTable setBackgroundColor:self.transparent];
    [self.resultTable setAllowsSelection:YES];
    [self.resultTable setUserInteractionEnabled:YES];
    [self addSubview:_resultTable];
    
    [self hideTable];
    
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    CGFloat padding = 5.0f;
    
    CGFloat x = padding;
    CGFloat y = Device.trueY0 + padding;
    CGFloat w = [self frame].size.width - 2 * padding;
    CGFloat h = 50;
    
    [self.inputField setFrame:CGRectMake(x, y, w, h)];
    
    y += h + 1;
    h = 240;
    
    [self.resultTable setFrame:CGRectMake(x, y, w, h)];
}


- (void)showTable {
    [self.resultTable setHidden:NO];
}

- (void)hideTable {
    [self.resultTable setHidden:YES];
}

- (void)closeTextField {
    [self.inputField resignFirstResponder];
    [self hideTable];
}

@end

File: AdvancedMap.Swift/Feature Demo/GeocodingView.swift
//
//  GeocodingView.swift
//  AdvancedMap.Swift
//
//  Created by Aare Undo on 06/07/2017.
//  Copyright © 2017 CARTO. All rights reserved.
//

import Foundation

class GeocodingView : BaseGeocodingView, UIGestureRecognizerDelegate {
    
    var inputField: UITextField!
    var resultTable: UITableView!
    
    let font = UIFont(name: "HelveticaNeue", size: 14)!
    
    let placeholder = "Type address..."

    convenience init() {
        self.init(frame: CGRect.zero)
        
        let attributes: [NSAttributedStringKey: AnyObject]? = [
            NSAttributedStringKey(rawValue: NSAttributedStringKey.foregroundColor.rawValue): Colors.fromRgba(red: 200, green: 200, blue: 200, alpha: 255),
            NSAttributedStringKey(rawValue: NSAttributedStringKey.font.rawValue): font
        ]
        
        initializeGeocodingView(popupTitle: Texts.geocodingInfoHeader, popupDescription: Texts.geocodingInfoContainer)

        inputField = UITextField()
        inputField.textColor = UIColor.white
        inputField.backgroundColor = Colors.darkTransparentGray
        inputField.autocorrectionType = .no
        inputField.font = font
        // Text padding
        inputField.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 10, height: 20))
        inputField.leftViewMode = .always
 
        inputField.attributedPlaceholder = NSAttributedString(string: placeholder, attributes: attributes)
        addSubview(inputField)
        
        resultTable = UITableView()
        resultTable.isHidden = true
        resultTable.register(GeocodingResultCell.self, forCellReuseIdentifier: GeocodingController.identifier)
        resultTable.backgroundColor = Colors.fromRgba(red: 0, green: 0, blue: 0, alpha: 0)
        resultTable.allowsSelection = true
        resultTable.isUserInteractionEnabled = true
        
        addSubview(resultTable)
    }
    
    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldReceive touch: UITouch) -> Bool {
        
        if (touch.view?.frame.contains(resultTable.frame))! {
            return true
        }
        
        return false
    }
    
    override func layoutSubviews() {
        
        super.layoutSubviews()
        
        let padding: CGFloat = 5
        let x: CGFloat = padding
        var y: CGFloat = Device.trueY0() + padding
        let w: CGFloat = frame.width - 2 * padding
        var h: CGFloat = 50
        
        inputField.frame = CGRect(x: x, y: y, width: w, height: h)
        
        y += h + 1
        h = 240

        resultTable.frame = CGRect(x: x, y: y, width: w, height: h)
    }
    
    override func addRecognizers() {
        super.addRecognizers()
        let recognizer = UITapGestureRecognizer(target: self, action: #selector(self.dismissKeyboard(_:)))
        recognizer.delegate = self
        self.addGestureRecognizer(recognizer)
    }
    
    override func removeRecognizers() {
        super.removeRecognizers()
    }
    
    @objc func dismissKeyboard(_ sender: UITapGestureRecognizer) {
        inputField.resignFirstResponder()
        resultTable.isHidden = true
    }
    
    func closeTextField() {
        inputField.resignFirstResponder()
        resultTable.isHidden = true
    }
    
    
    func showBannerInsteadOfSearchBar() {
        inputField.isHidden = true
        banner.show(text: "DOWNLOAD A PACKAGE TO START GEOCODING")
    }
    
    func showSearchBar() {
        inputField.isHidden = false
        banner.hide()
    }
}










